clippy_lints/src/assert_in_result_fn.rs

```diff
@@ -0,0 +1,73 @@
+use crate::utils::{find_macro_calls, is_type_diagnostic_item, return_ty, span_lint_and_then};
+use rustc_hir::intravisit::FnKind;
+use rustc_hir::{Body, FnDecl, HirId};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::Span;
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for usage of `assert!()`, `assert_eq!()` or `assert_ne!()` in functions returning a `Result`.
+    ///
+    /// **Why is this bad?** For some codebases, it is desirable for functions of return type `Result` to return an `Err` instead of crashing. Hence assertions should be avoided.
+    ///
+    /// **Known problems:** Functions not returning a `Result` may be called by a function returning a `Result`, and these functions may contain assertions. This is not checked.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// fn divide(num: isize, den: isize) -> Result<f64, String> {
+    ///     assert_ne!(den, 0, "division by zero");
+    ///     Ok(num as f64 / den as f64)
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// fn divide(num: isize, den: isize) -> Result<f64, String> {
+    ///     if den == 0 {
+    ///         return Err(String::from("division by zero"));
+    ///     }
+    ///     Ok(num as f64 / den as f64)
+    /// }
+    /// ```
+    pub ASSERT_IN_RESULT_FN,
+    restriction,
+    "functions of type `Result<..>` that contain `assert!()`, `assert_eq!()` or `assert_ne!()`"
+}
+
+declare_lint_pass!(AssertInResultFn => [ASSERT_IN_RESULT_FN]);
+
+impl LateLintPass<'tcx> for AssertInResultFn {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'tcx>,
+        fn_kind: FnKind<'tcx>,
+        _: &'tcx FnDecl<'tcx>,
+        body: &'tcx Body<'tcx>,
+        span: Span,
+        hir_id: HirId,
+    ) {
+        if !matches!(fn_kind, FnKind::Closure(_))
+            && is_type_diagnostic_item(cx, return_ty(cx, hir_id), sym!(result_type))
+        {
+            lint_impl_body(cx, span, body);
+        }
+    }
+}
+
+fn lint_impl_body<'tcx>(cx: &LateContext<'tcx>, impl_span: Span, body: &'tcx Body<'tcx>) {
+    let asserts = find_macro_calls(vec!["assert", "assert_eq", "assert_ne"], body);
```

Comment:
We should also check for the debug variants of these macros: `debug_assert`, `debug_assert_eq` and `debug_assert_ne`.

Reply:
Done in new commits.